### PR TITLE
Add KeyedFileMutex

### PR DESCRIPTION
**Pull request by bwoebi**

### CODE REVIEW

File: test/Sync/AsyncFileKeyedMutexTest.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $this->ignoreLoopWatchers();
+    }
+
+    public function createMutex(): KeyedMutex
+    {
+        return new AsyncFileKeyedMutex(\tempnam(\sys_get_temp_dir(), 'mutex-') . '%s.lock');
```
Comment (kelunik) [inline test/Sync/AsyncFileKeyedMutexTest.php:20]: tempnam shouldn't be used here, as it creates an additional file that isn't used. I guess a bit of random will be fine.

File: src/Sync/AsyncFileKeyedMutex.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+use Amp\File\FilesystemException;
+use Amp\Sync\KeyedMutex;
+use Amp\Sync\Lock;
+use Amp\Sync\SyncException;
+use function Amp\delay;
+use function Amp\File\openFile;
+use function Amp\File\deleteFile;
+
+final class AsyncFileKeyedMutex implements KeyedMutex
```
Comment (kelunik) [inline src/Sync/AsyncFileKeyedMutex.php:13]: I think we should drop `Async` from the name and name it `KeyedFileMutex`. We should introduce a class alias for the existing implementation without the `Async`prefix as well.

Comment (bwoebi) [reply src/Sync/AsyncFileKeyedMutex.php:13]: I agree, and I think we do not need a class alias, given the new major anyway

Comment (kelunik) [reply src/Sync/AsyncFileKeyedMutex.php]: Good point. I'd still switch the order of words and name it `KeyedFileMutex` instead of `FileKeyedMutex`.

Comment (bwoebi) [reply src/Sync/AsyncFileKeyedMutex.php]: True, that's consistent with amphp/sync, so, done.

File: src/Sync/KeyedFileMutex.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    private const LATENCY_TIMEOUT = 10;
+
+    private string $pattern;
+
+    /**
+     * @param string $pattern Name of temporary file to use as a mutex, including a %s to splice the key in
+     */
+    public function __construct(string $pattern)
+    {
+        if (!preg_match("((%%(*SKIP))*%s)", $pattern)) {
```
Comment (kelunik) [inline src/Sync/KeyedFileMutex.php:24]: Any reason to still use `%s` instead of making `$pattern` a prefix?

File: src/KeyedFileMutex.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                $file = $this->filesystem->openFile($filename, 'x');
+
+                // Return a lock object that can be used to release the lock on the mutex.
+                $lock = new Lock(fn () => $this->release($filename));
+
+                $file->close();
+
+                return $lock;
+            } catch (FilesystemException) {
+                delay(self::LATENCY_TIMEOUT);
```
Comment (kelunik) [inline src/KeyedFileMutex.php:49]: For longer held locks, this is very inefficient. Should we backoff exponentially up to 1 second?

Comment (trowski) [reply src/KeyedFileMutex.php:49]: Probably a good idea. This is exactly what we're doing for `FileMutex` as well, so I'd say let's update both then.

**Pull request merged**